protocol/nats_jetstream/v3/message_test.go

```diff
@@ -0,0 +1,116 @@
+ SPDX-License-Identifier: Apache-2.0
+*/
+
+package nats_jetstream
+
+import (
+	"context"
+	"encoding/json"
+	"testing"
+
+	"github.com/cloudevents/sdk-go/v2/binding/spec"
+	bindingtest "github.com/cloudevents/sdk-go/v2/binding/test"
+
+	"github.com/cloudevents/sdk-go/v2/binding"
+	"github.com/cloudevents/sdk-go/v2/test"
+	"github.com/nats-io/nats.go"
+	"github.com/nats-io/nats.go/jetstream"
+)
+
+type jetStreamMsg struct {
+	jetstream.Msg
+	msg *nats.Msg
+}
+
+func (j *jetStreamMsg) Data() []byte         { return j.msg.Data }
+func (j *jetStreamMsg) Headers() nats.Header { return j.msg.Header }
+
+var (
+	outBinaryMessage = bindingtest.MockBinaryMessage{
+		Metadata:   map[spec.Attribute]interface{}{},
+		Extensions: map[string]interface{}{},
+	}
+	outStructMessage = bindingtest.MockStructuredMessage{}
+
+	testEvent     = test.FullEvent()
+	binaryData, _ = json.Marshal(map[string]string{
+		"ce-type":            testEvent.Type(),
+		"ce-source":          testEvent.Source(),
+		"ce-id":              testEvent.ID(),
+		"ce-time":            test.Timestamp.String(),
+		"ce-specversion":     "1.0",
+		"ce-dataschema":      test.Schema.String(),
+		"ce-datacontenttype": "text/json",
+		"ce-subject":         "receiverTopic",
+		"ce-exta":            "someext",
+	})
+	structuredReceiverMessage = &jetStreamMsg{
+		msg: &nats.Msg{
+			Subject: "hello",
+			Data:    binaryData,
+		},
+	}
+	binaryReceiverMessage = &jetStreamMsg{
+		msg: &nats.Msg{
+			Subject: "hello",
+			Data:    testEvent.Data(),
+			Header: nats.Header{
+				"ce-type":            {testEvent.Type()},
+				"ce-source":          {testEvent.Source()},
+				"ce-id":              {testEvent.ID()},
+				"ce-time":            {test.Timestamp.String()},
+				"ce-specversion":     {"1.0"},
+				"ce-dataschema":      {test.Schema.String()},
+				"ce-datacontenttype": {"text/json"},
+				"ce-subject":         {"receiverTopic"},
+				"ce-exta":            {"someext"},
+			},
+		},
+	}
+)
+
+func TestNewMessage(t *testing.T) {
+	tests := []struct {
+		name                    string
+		receiverMessage         jetstream.Msg
+		expectedEncoding        binding.Encoding
+		expectedStructuredError error
+		expectedBinaryError     error
+	}{
+		{
+			name:                    "Structured encoding",
+			receiverMessage:         structuredReceiverMessage,
+			expectedEncoding:        binding.EncodingStructured,
+			expectedStructuredError: nil,
+			expectedBinaryError:     binding.ErrNotBinary,
+		},
+		{
+			name:                    "Binary encoding",
+			receiverMessage:         binaryReceiverMessage,
+			expectedEncoding:        binding.EncodingBinary,
+			expectedStructuredError: binding.ErrNotStructured,
+			expectedBinaryError:     nil,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			got := NewMessage(tt.receiverMessage)
+			if got == nil {
+				t.Errorf("Error in NewMessage!")
+			}
+			err := got.ReadBinary(context.TODO(), &outBinaryMessage)
```

Comment:
nit: shouldn't you also assert that the contents of the messages are as expected?

Reply:
done